src/main/java/org/gaul/s3proxy/junit/S3ProxyRule.java

```diff
@@ -0,0 +1,189 @@
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.BlobStoreContext;
+import org.jclouds.blobstore.domain.StorageMetadata;
+import org.junit.rules.ExternalResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.Beta;
+
+/**
+ * A JUnit Rule that manages an S3Proxy instance which tests can use as an S3 API endpoint.
+ */
+@Beta
+public final class S3ProxyRule extends ExternalResource {
+    private static final Logger logger = LoggerFactory.getLogger(
+        S3ProxyRule.class);
+
+    private static final String LOCALHOST = "127.0.0.1";
+
+    private final String accessKey;
+    private final String secretKey;
+    private final String endpointFormat;
+    private final S3Proxy s3Proxy;
+
+    private BlobStoreContext blobStoreContext;
+    private URI endpointUri;
+
+    public static final class Builder {
+        private AuthenticationType authType = AuthenticationType.NONE;
+        private String accessKey;
+        private String secretKey;
+        private String secretStorePath;
+        private String secretStorePassword;
+        private int port = -1;
+        private boolean ignoreUnknownHeaders;
+        private String blobStoreProvider = "filesystem";
+
+        private Builder() { }
+
+        public Builder withCredentials(AuthenticationType authType,
+                String accessKey, String secretKey) {
+            this.authType = authType;
+            this.accessKey = accessKey;
+            this.secretKey = secretKey;
+            return this;
+        }
+
+        public Builder withCredentials(String accessKey, String secretKey) {
+            return withCredentials(AuthenticationType.AWS_V2_OR_V4, accessKey,
+                secretKey);
+        }
+
+        public Builder withSecretStore(String path, String password) {
+            secretStorePath = path;
+            secretStorePassword = password;
+            return this;
+        }
+
+        public Builder withPort(int port) {
+            this.port = port;
+            return this;
+        }
+
+        public Builder withBlobStoreProvider(String blobStoreProvider) {
+            this.blobStoreProvider = blobStoreProvider;
+            return this;
+        }
+
+        public Builder ignoreUnknownHeaders() {
+            ignoreUnknownHeaders = true;
+            return this;
+        }
+
+        public S3ProxyRule build() {
+            return new S3ProxyRule(this);
+        }
+    }
+
+    private S3ProxyRule(Builder builder) {
+        accessKey = builder.accessKey;
+        secretKey = builder.secretKey;
+
+        Properties properties = new Properties();
+        try {
+            File blobStoreLocation = Files.createTempDirectory("S3ProxyRule")
+                    .toFile();
+            blobStoreLocation.deleteOnExit();
+            properties.setProperty("jclouds.filesystem.basedir",
+                blobStoreLocation.getCanonicalPath());
+        } catch (IOException e) {
```

Comment:
You can narrow the exception scope by declaring `File` outside `try` and initializing it inside.

Reply:
Not sure I get you. This will then just have an extra line

    File blobStoreLocation;

before the try? This unnecessarily increases the scope of `blobStoreLocation`. Or did I misunderstand?

Reply:
Sorry I was not clear, moving the `File` declaration allows you to move the `setProperty` call out of the exception scope.

Reply:
OK, I can do that, it increases the scope of the File but fine.

Reply:
Actually, no, `blobStoreLocation.getCanonicalPath()` can also throw an IOException...